### PR TITLE
Host resources wit-bindgen code generation

### PR DESCRIPTION
**Pull request by silesmo**
Implements host export resource generation #6722 

After a discussion with @alexcrichton, I have broken out basic resource generation for host resource exports into it's own PR,
Host import resource generation will follow, as well as more ergonomic resource code generation that abstracts away the raw resource handles.

Example implementation: https://github.com/Nor2-io/wasmtime-resources-example/blob/main/example-host-export-index/src/main.rs

Example code generation output: https://github.com/Nor2-io/wasmtime-resources-example/blob/main/example-host-export-index/main%20copy%202.rs

Sneak peak of future additions that abstract away resources can be seen in the link below. (It doesn't work with nested resources yet.)
https://github.com/Nor2-io/wasmtime-resources-example/blob/main/example-host-import/src/main.rs
https://github.com/Nor2-io/wasmtime-resources-example/blob/main/example-host-export/src/main.rs

Added temporary exception for cargo deny on multiple versions until memfd has been updated: [#62](https://github.com/lucab/memfd-rs/pull/62)

### CODE REVIEW

File: crates/wasi-http/Cargo.toml [hunk 1]
```diff
@@ -31,4 +31,4 @@ wasmtime = { workspace = true, features = ['component-model'] }
 [target.'cfg(not(any(target_arch = "riscv64", target_arch = "s390x")))'.dependencies]
 tokio-rustls = { version = "0.24.0" }
 rustls = { version = "0.21.0" }
-webpki-roots = { version = "0.23.0" }
+webpki-roots = { version = "0.25.2" }
```
Comment (alexcrichton) [inline crates/wasi-http/Cargo.toml:34]: I think this (and the other `wasi-http` changes) could be separated out from this?

**Review by alexcrichton — COMMENTED** [review]
Thanks again for this! I'm running out of time today but wanted to give some quick feedback on this. There's a few comments below and I have more thoughts on the generated code but I'll try to flesh that out more tomorrow (I'll probably send a PR-to-your-PR in the morning)

In the meantime though this'll need to have some tests as well, both at `crates/component-macro/tests/codegen/*.wit` and additionally one in `tests/all/component_model/*`. You'll find some other bindgen-related tests in that directory too.

File: crates/wasi-http/Cargo.toml (hunk 1 above)
Comment (silesmo) [reply crates/wasi-http/Cargo.toml:34]: Yes, If I add an exception for cargo deny I guess? I just made the updates to not have duplicated dependencies to pass Cargo deny. Is that the way to go and then a PR that makes the change and removes the exception or? I haven't used cargo deny in a work flow before so I appreciate guidance on this 🙂

File: crates/wasmtime/src/component/linker.rs [hunk 2]
```diff
@@ -399,14 +399,12 @@ impl<T> LinkerInstance<'_, T> {
     pub fn resource<U: 'static>(
         &mut self,
         name: &str,
-        dtor: impl Fn(StoreContextMut<'_, T>, u32) + Send + Sync + 'static,
+        dtor: impl Fn(StoreContextMut<'_, T>, u32) -> Result<()> + Send + Sync + 'static,
```
[the remark this answers is not in the document]
Comment (silesmo) [reply crates/wasmtime/src/component/linker.rs:402]: Will do!

File: Cargo.toml [hunk 3]
```diff
@@ -51,14 +51,24 @@ rustix = { workspace = true, features = ["mm", "param"] }
 
 [dev-dependencies]
 # depend again on wasmtime to activate its default features for tests
-wasmtime = { workspace = true, features = ['component-model', 'async', 'default', 'winch'] }
+wasmtime = { workspace = true, features = [
+  'component-model',
```
[the remark this answers is not in the document]
Comment (silesmo) [reply Cargo.toml:55]: Yes will revert.

File: Cargo.lock [hunk 4]
```diff
@@ -19,9 +19,9 @@ checksum = "f26201604c87b1e01bd3d98f8d5d9a8fcbb815e8cedb41ffccbeb4bf593a35fe"
 
 [[package]]
 name = "ahash"
-version = "0.8.2"
+version = "0.8.3"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bf6ccdb167abbf410dcb915cabd428929d7f6a04980b54a11f26a39f1c7f7107"
+checksum = "2c99f64d1e06488f620f932677e24bc6e2897582980441ae90a671415bd7ec2f"
```
[the remark this answers is not in the document]
Comment (silesmo) [reply Cargo.lock:24]: I will take a look.

File: crates/wasmtime/src/component/linker.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (silesmo) [reply crates/wasmtime/src/component/linker.rs:408]: added doc comment

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (silesmo) [reply Cargo.toml]: done

File: crates/wasi-http/Cargo.toml (hunk 1 above)
Comment (silesmo) [reply crates/wasi-http/Cargo.toml]: Moved into it's own PR: #6886

File: Cargo.lock (hunk 4 above)
[the remark this answers is not in the document]
Comment (silesmo) [reply Cargo.lock:24]: removed the updates, so now the cargo-deny errors are back.

**Comment by alexcrichton** [comment]
Ok I've worked on this a bit this morning and I've pushed up https://github.com/Nor2-io/wasmtime/pull/1 which contains another round of comments for me (and some fixes for the deny/vet-related things I think). As I mentioned over there I'm happy to go through anything if you'd like clarifications or explanations!

**Comment by alexcrichton** [comment]
One more round of changes at https://github.com/Nor2-io/wasmtime/pull/2 (again happy to go through anything if you'd like), and then I think this is good to go 👍

**Review by alexcrichton — APPROVED** [review]
Thanks again @silesmo for all your work on this!

Since this now has bits and pieces from me too I'm going to ask @pchickey to do a once-over on this to confirm it's all reasonable, but otherwise I think this is good to go 👍

**Comment by alexcrichton** [comment]
Also @silesmo would you be up for filing an issue for follow-up work?

**Comment by silesmo** [comment]
> Also @silesmo would you be up for filing an issue for follow-up work?

Yes I can do that! 🙂

**Comment by alexcrichton** [comment]
Pat's busy with streams and such so he'll take a look at this after-the-fact, so I'm going to go ahead and flag this for merge. Thanks again @silesmo!

**Comment by alexcrichton** [comment]
oh I think this may need a rebase for merging now?

**Comment by silesmo** [comment]
> oh I think this may need a rebase for merging now?

I merged the latest changes now

**Pull request merged**